### PR TITLE
fix: minor code correction

### PR DESCRIPTION
**Pull request by gbaranski**

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -1,73 +1,83 @@
...
-            let input = input.trim().to_string();
-
-            if input.to_lowercase() == hidden_word.to_lowercase() {
-                good_answers += 1;
-                stdout.queue(SetForegroundColor(Color::Green)).unwrap();
-                print!("Good!");
+    lines
+        .iter()
+        .enumerate()
+        .fold(0, |correct_answers, (i, line)| {
```
Comment (jelni) [inline src/main.rs:29]: Not sure if it's a common Rust practice, but I find using `fold` to keep track of `correct_answers` hard to read

File: src/main.rs [hunk 2]
```diff
@@ -1,73 +1,83 @@
...
+            // clear_console(&mut stdout);
+            let (line, hidden_word) = hide_word(line);
+            stdout.queue(Print(line.blue())).unwrap();
+            stdout.queue(Print("> ")).unwrap();
+            stdout.flush().unwrap();
+            let input = stdin.next().unwrap().unwrap();
+            let input = input.trim();
+            let is_valid = input.eq_ignore_ascii_case(&hidden_word);
```
Comment (jelni) [inline src/main.rs]: `str1.eq_ignore_ascii_case(str2)` is not the same as `str1.to_lowercase() == str2.to_lowercase()` (it only compares ASCII chars)

File: src/main.rs [hunk 3]
```diff
@@ -1,73 +1,83 @@
...
+            let (correct_answers, foreground_color, message) = if is_valid {
+                (
+                    correct_answers + 1,
+                    Color::Green,
+                    format!("Correct answer!"),
```
Comment (jelni) [inline src/main.rs]: Unnecessary `format!` usage

File: src/main.rs [hunk 4]
```diff
@@ -1,73 +1,83 @@
...
-    let word_count = words.clone().count();
-    let mut words: Vec<&str> = words.collect();
+// here we're using AsRef<str> instead of &str or String, because AsRef<str> can accept both of these.
+fn hide_word(line: impl AsRef<str>) -> (String, String) {
+    let line = line.as_ref();
+    let words = line.split_whitespace().collect::<Vec<_>>();
+    let n = rand::thread_rng().gen_range(0..words.len());
+    let hidden_word = words.get(n).unwrap().to_string();
+    // eventually replace it with some constant if you want to make it harder for user to recognize the hidden word
+    let censored_word = "_".repeat(hidden_word.len());
```
Comment (jelni) [inline src/main.rs]: I would prefer it to always be `"___"`

File: src/main.rs [hunk 5]
```diff
@@ -1,73 +1,83 @@
...
 
-    let index = rand.gen_range(0..word_count);
-    let word = words[index];
-    let captures = WORD_RE.captures(word).unwrap();
-    let hidden_word: String = captures.name("word").unwrap().as_str().to_string();
```
Comment (jelni) [inline src/main.rs:62]: This regex clears punctuation marks from the word (like `,.?!`) and you shouldn't remove it

**Review by jelni — CHANGES_REQUESTED** [review]
Thank you for the PR! Just a few things I'm not sure about

File: src/main.rs [hunk 6]
```diff
@@ -1,73 +1,83 @@
...
-    word.push_str(captures.name("start").unwrap().as_str());
-    word.push_str("___");
-    word.push_str(captures.name("end").unwrap().as_str());
-    words[index] = word.as_str();
-    let new_line = words.join(" ");
+    // encapsulate mutable access into a block, so that it's clearly visible where it gets mutated.
+    let words = {
+        let mut words = words;
+        words[n] = &censored_word;
+        words
```
Comment (jelni) [inline src/main.rs]: This feels very over-engineered

File: src/main.rs (hunk 1 above)
Comment (gbaranski) [reply src/main.rs:29]: that's the more functional or declarative way of programming, Iterators are pretty common in Rust world.

File: src/main.rs [hunk 7]
```diff
@@ -12,37 +12,37 @@ fn main() {
...
         .fold(0, |correct_answers, (i, line)| {
-            // clear_console(&mut stdout);
             let (line, hidden_word) = hide_word(line);
             stdout.queue(Print(line.blue())).unwrap();
             stdout.queue(Print("> ")).unwrap();
             stdout.flush().unwrap();
             let input = stdin.next().unwrap().unwrap();
             let input = input.trim();
-            let is_valid = input.eq_ignore_ascii_case(&hidden_word);
+            let is_valid = input.to_lowercase() == hidden_word.to_lowercase();
```
Comment (gbaranski) [inline src/main.rs:35]: why?

File: src/main.rs [hunk 8]
```diff
@@ -63,21 +63,12 @@ fn main() {
 // here we're using AsRef<str> instead of &str or String, because AsRef<str> can accept both of these.
 fn hide_word(line: impl AsRef<str>) -> (String, String) {
```
Comment (gbaranski) [inline src/main.rs:57]: why?

File: src/main.rs (hunk 2 above)
Comment (gbaranski) [reply src/main.rs]: well yes, that's on purpose, some people are too lazy to use diacritics and I wouldn't treat this as a mistake.

File: src/main.rs (hunk 3 above)
Comment (gbaranski) [reply src/main.rs]: yes, but it looks more "symmetrical" because few lines below you've got the similar `format!()` call

File: src/main.rs (hunk 6 above)
Comment (gbaranski) [reply src/main.rs]: well, maybe on this example that could be called "over-engineer", but it's important to follow good code practices everywhere.

File: src/main.rs (hunk 2 above)
Comment (michalwa) [reply src/main.rs]: `eq_ignore_ascii_case` doesn't ignore diacritics

**Review by jelni — APPROVED** [review]

**Pull request merged**